Read next prev on

hsribei opened the pull request:
Functionality working(-ish). Previous/Next links lead to their respective sibling nodes, and they use `GatsbyLink` to make client-side navigation and prefetching amazing. Probably not the prettiest so far. I'm opening the PR so we can iterate on that.

Aaand... I just realized the order is wrong. GraphQL returns markdown nodes ordered by filename, and they are not left-padded, so after `1.md` comes `11.md` instead of `2.md`. If it's an acceptable solution, I will rename all files to `01.md`, etc. then check if nothing else breaks.

Also, the `/react/` articles come before the `/state/` articles, as they should, but that is only a coincidental feature of their lexicographic order. When you add a new module like `/build-a-full-stack-app/`, if it should come after `/state/`, it won't, so we need to come up with a way to fix this. Maybe a `/modules/` directory with pages for each and hardcoded prev/next slug fields in the frontmatter?

Relevant commit:

```diff
commit 9b654167d82a3e175cefa1b98d915b07d41851ce
Author: Helder S Ribeiro <hsribei.pub@gmail.com>
Date:   Mon Jul 9 21:28:48 2018 -0400

    WIP Add previou/next pagination for articles
---
 gatsby-node.js           | 12 ++++++++++--
 src/templates/article.js | 26 +++++++++++++++++++++++---
 2 files changed, 33 insertions(+), 5 deletions(-)

diff --git a/gatsby-node.js b/gatsby-node.js
index 4df1304..0e5ddf9 100644
--- a/gatsby-node.js
+++ b/gatsby-node.js
@@ -114,15 +114,23 @@ exports.createPages = ({ graphql, actions }) => {
         }
       }
     `).then(result => {
-      result.data.allMarkdownRemark.edges.forEach(({ node }) => {
+      result.data.allMarkdownRemark.edges.forEach(
+        ({ node }, index, articles) => {
           createPage({
             path: node.fields.slug,
             component: path.resolve('./src/templates/article.js'),
             context: {
               slug: node.fields.slug,
+              prevSlug:
+                index === 0 ? null : articles[index - 1].node.fields.slug,
+              nextSlug:
+                index === articles.length - 1
+                  ? null
+                  : articles[index + 1].node.fields.slug,
             },
           })
-      })
+        }
+      )
       resolve()
     })
   })
diff --git a/src/templates/article.js b/src/templates/article.js
index ffe927d..c5bcdc8 100644
--- a/src/templates/article.js
+++ b/src/templates/article.js
@@ -3,11 +3,14 @@ import React from 'react'
 import Layout from '../components/layout'
 import ResponsivePlayer from '../components/ResponsivePlayer'
 
-import { LowSection, SectionTitle } from '../components/Section'
-import { Row, Col as Column } from 'react-bootstrap'
+import { LowSection } from '../components/Section'
+import { Pager, Row, Col as Column } from 'react-bootstrap'
 
-export default ({ data }) => {
+import { Link } from 'gatsby'
+
+export default ({ data, pathContext }) => {
   const article = data.markdownRemark
+  const { prevSlug, nextSlug } = pathContext
 
   return (
     <Layout>
@@ -21,6 +24,23 @@ export default ({ data }) => {
           </Column>
         </Row>
       </LowSection>
+      <Pager>
+        <p>👓&nbsp;Keep on learnin'&nbsp;👓</p>
+        <Pager.Item
+          href={'/' + prevSlug}
+          componentClass={Link}
+          to={'/' + prevSlug}
+        >
+          &larr;&nbsp;Previous article
+        </Pager.Item>
+        <Pager.Item
+          href={'/' + nextSlug}
+          componentClass={Link}
+          to={'/' + nextSlug}
+        >
+          Next article&nbsp;&rarr;
+        </Pager.Item>
+      </Pager>
     </Layout>
   )
 }
```

hsribei committed:
Remove unmet and unneeded dependency

Fixes build. `gatsby-plugin-react-next` is no longer necessary
as we're already using gatsby v2 which uses React 16.

hsribei committed:
Fix build. There's no /exercises directory.

hsribei committed:
Fix client_secrets.json key name

hsribei committed:
Fix videos not loading on landing page

hsribei committed:
WIP Add previou/next pagination for articles

hsribei committed:
Add `npm run debug-build` for easier debugging

Swizec commented on gatsby-config.js:
```diff
@@ -5,7 +5,6 @@ module.exports = {
       '👉 No pressure learning at 2 minutes per day. Start with the basics,become a React 16.3 master',
   },
   plugins: [
-    'gatsby-plugin-react-next',
```
Did I forget to remove this when upgrading to Gatsby v2?

hsribei replied:
yep

Swizec commented on gatsby-config.js:
```diff
@@ -50,20 +49,6 @@ module.exports = {
             },
           },
-          {
-            resolve: `gatsby-remark-code-repls`,
```
Why are we removing this? It's a useful plugin that I intend to use in some of the articles.

If it's because the `examples` directory is missing, a better approach is to make that directory and put a `.keep` file in it so git doesn't drop it due to the empty directory optimization.

hsribei replied:
Ah ok, removed it as it was unused and was breaking, but will add it back. There are some direct embeds from codesandbox that i could also port to it if you want me to.

Swizec replied:
We can deal with those later. I wanna find a way to have them as embeds, but maintained through an examples dir.

Swizec commented on gatsby-node.js:
```diff
@@ -114,15 +114,23 @@ exports.createPages = ({ graphql, actions }) => {
+            context: {
+              slug: node.fields.slug,
+              prevSlug:
+                index === 0 ? null : articles[index - 1].node.fields.slug,
```
What if index is less than 0 because of a bug or hack?

A better approach in these cases is to check that the thing you're looking for exists, like this

```javascript
prevSlug: articles[index - 1] && articles[index - 1].node.fields.slug
```

This returns `null` because of [short circuit evaluation](https://en.wikipedia.org/wiki/Short-circuit_evaluation), or the value you want if possible.

hsribei replied:
Good defensiveness trick, will change to that.

Swizec commented on gatsby-node.js:
```diff
@@ -114,15 +114,23 @@ exports.createPages = ({ graphql, actions }) => {
+              prevSlug:
+                index === 0 ? null : articles[index - 1].node.fields.slug,
+              nextSlug:
+                index === articles.length - 1
```
Same comment as above, what if your index jumps beyond the array edge?

Swizec commented on package.json:
```diff
@@ -41,6 +40,7 @@
     "scripts": {
         "deploy": "gatsby build && cd public/ && now && cd ..",
         "build": "gatsby build",
+        "debug-build": "node --inspect-brk ./node_modules/gatsby/dist/bin/gatsby.js build",
```
Neat!

Swizec commented on src/components/CurriculumEntry.js:
```diff
@@ -40,7 +40,7 @@ class CurriculumEntry extends React.Component {
         {typeof window === 'undefined' ? null : (
           <AnimakitExpander expanded={expanded}>
-            <ResponsivePlayer videoId={video.videId} />
+            <ResponsivePlayer videoId={video.videoId} />
```
Ha, I didn't even notice that!

hsribei replied:
Odd thing is that this was making the landing page videos not load at all on my localhost, but on production they were loading fine 🤔

Swizec commented on src/templates/article.js:
```diff
@@ -21,6 +24,23 @@ export default ({ data }) => {
         </Row>
       </LowSection>
+      <Pager>
+        <p>👓&nbsp;Keep on learnin'&nbsp;👓</p>
```
Love the emoji flare :)

hsribei replied:
Keeping with the spirit =)

Swizec commented on src/templates/article.js:
```diff
@@ -21,6 +24,23 @@ export default ({ data }) => {
+          componentClass={Link}
+          to={'/' + prevSlug}
+        >
+          &larr;&nbsp;Previous article
```
Let's render article titles here as well. Makes it more clicky.

(we're gonna add thumbnails in a separate ticket/PR)

Swizec replied:
Also let's call them lessons

hsribei replied:
So the button label should be like "Previous lesson: How to MobX in a Nutshell"?

Swizec commented:
Looks like clicking "previous" on the first lesson in the MobX series puts me here

<img width="309" alt="screen shot 2018-07-09 at 23 30 00" src="https://user-images.githubusercontent.com/56883/42493032-1c7cd094-83d0-11e8-81e4-39f5eca8fffd.png">

Let's fix that :)

hsribei committed:
Revert "Fix build. There's no /exercises directory."

This reverts commit 6911ca59e9426d718a0f4db3ac16d050d0e315da.

hsribei committed:
Fixes build w/o removing gatsby-remark-code-repls

ht @swizec

hsribei committed:
Zero-pad md articles so filepath sort order works

This was done using this amazing gist by yours truly:

https://gist.github.com/hsribei/13dfd3ae45df0f9881b1567ee37bcd8a

hsribei committed:
Fix inter- and intra-module pagination order

- Between modules, `/state/` articles were coming before `react` ones
for some odd reason related to how `gatsby-source-filesystem` loads
files into graphql.
- Within modules, order was like `1.md`, `10.md`, `2.md` because gatsby
was loading files in lexicographic order. It wasn't *exactly* in that
order either, because no `sort` criterium was being passed, so it was
almost random.

Renaming the files to `01.md`, etc. (left zero padding, previous commit)
and using that as the order (this commit) solved both problems.

hsribei committed:
Fix prev/next pagination links on first/last pages

They were pointing to /null. Now they are not displayed.

hsribei committed:
Make array indexing defensive with short-circuiting

hsribei committed:
Rename s/article/lesson/ & add title to pagination

hsribei commented:
Should be all fixed!

Swizec commented:
Found a bug, this page doesn't recognize that there's a next page: http://localhost:8000/state/how-to-mobx-in-a-nutshell

hsribei commented:
I can't reproduce this, are you still observing it? Can you send me the
sha1 you tested it with plz? Thanks!

On Thu, Jul 12, 2018 at 4:53 AM Swizec Teller <notifications@github.com>
wrote:

> Found a bug, this page doesn't recognize that there's a next page:
> http://localhost:8000/state/how-to-mobx-in-a-nutshell
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Swizec/learnwhileyoupoop.com/pull/4#issuecomment-404440205>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAQ238_nzYU7sorLvLGn-sIoE589O-_ks5uFw6lgaJpZM4VIsRw>
> .
>
